src/corefx/System.Globalization.Native/locale.cpp

```diff
@@ -0,0 +1,373 @@
+	u_charsToUChars(str, value, len + 1);
+	return U_ZERO_ERROR;
+}
+
+/*
+PAL Function:
+getCanonicalLocaleName
+
+Obtains a canonical locale given the locale name
+*/
+extern "C" int32_t GetLocaleName(const UChar* localeName, UChar* value, int32_t valueLength)
+{
+	Locale locale = GetLocale(localeName, true);
+
+	if (locale.isBogus())
+	{
+		return U_ILLEGAL_ARGUMENT_ERROR;
+	}
+
+	if (strlen(locale.getISO3Language()) == 0)
+	{
+		// unknown language; language is required (script and country optional)
+		return U_ILLEGAL_ARGUMENT_ERROR;
+	}
+
+	UErrorCode status = CopyCharToUChar(locale.getName(), value, valueLength);
+	if (U_SUCCESS(status))
+	{
+		// replace underscores with hyphens to interop with existing .NET code
+		for (UChar* ch = value; *ch != (UChar)'\0'; ch++)
+		{
+			if (*ch == (UChar)'_')
+			{
+				*ch = (UChar)'-';
+			}
+		}
+	}
+
+	return status;
+}
+
+/*
+Function:
+GetMeasurementSystem
+
+Obtains the measurement system for the local, determining if US or metric
+*/
+UErrorCode GetMeasurementSystem(const char *localeId, int32_t *value)
+{
+	UErrorCode status = U_ZERO_ERROR;
+
+	UMeasurementSystem measurementSystem = ulocdata_getMeasurementSystem(localeId, &status);
+	if (U_SUCCESS(status))
+	{
+		*value = (measurementSystem == UMeasurementSystem::UMS_US) ? 1 : 0;
+	}
+
+	return status;
+}
+
+/*
+PAL Function:
+GetLocaleInfoInt
+
+Returns integer locale information
+*/
+extern "C" int32_t GetLocaleInfoInt(const UChar* localeName, LocaleNumberData localeNumberData, int32_t* value)
+{
+	Locale locale = GetLocale(localeName);
+	if (locale.isBogus())
+	{
+		return U_ILLEGAL_ARGUMENT_ERROR;
+	}
+
+	UErrorCode status = U_ZERO_ERROR;
+	switch (localeNumberData)
+	{
+		case LanguageId:
+			*value = locale.getLCID();
+			break;
+		case MeasurementSystem:
+			status = GetMeasurementSystem(locale.getName(), value);
+			break;
+		case FractionalDigitsCount:
+			*value = 2; //TODO: implement
+			break;
+		case NegativeNumberFormat:
+			*value = 1; //TODO: implement
+			break;
+		case MonetaryFractionalDigitsCount:
+			*value = 2; //TODO: implement
+			break;
+		case PositiveMonetaryNumberFormat:
+			*value = 0; //TODO: implement
+			break;
+		case NegativeMonetaryNumberFormat:
+			*value = 0; //TODO: implement
+			break;
+		case CalendarType:
+			*value = 1; //TODO: implement
+			break;
+		case FirstWeekOfYear:
+			*value = 0; //TODO: implement
+			break;
+		case ReadingLayout:
+			*value = 0; //TODO: implement
+			break;
+		case NegativePercentFormat:
+			*value = 0; //TODO: implement
+			break;
+		case PositivePercentFormat:
+			*value = 0; //TODO: implement
+			break;
+		default:
+			status = U_UNSUPPORTED_ERROR;
+			break;
+	}
+
+	return status;
+}
+
+/*
+Function:
+GetLocaleInfoDecimalFormatSymbol
+
+Obtains the value of a DecimalFormatSymbols
+*/
+UErrorCode GetLocaleInfoDecimalFormatSymbol(const Locale &locale, DecimalFormatSymbols::ENumberFormatSymbol symbol, UChar* value, int32_t valueLength)
+{
+	UErrorCode status = U_ZERO_ERROR;
```

Comment:
Feels like there is a lot of ceremony in these methods (to create the object, check for errors, etc).  We seem to have to do it a lot.  Is there a way to factor this into something where we have less duplicated code?  Are things better if we just used the C API which might hide a lot of this from us?